vision_agent/agent/agent_coder.py

```diff
@@ -0,0 +1,170 @@
+) -> Dict[str, Union[str, bool]]:
+    prompt = VISUAL_TEST.format(
+        docstring=TOOLS_DOCSTRING,
+        code=code,
+        question=question,
+        feedback=feedback,
+    )
+    completion = model(prompt, images=[viz_file])
+    # type is from the prompt
+    return json.loads(completion)  # type: ignore
+
+
+def fix_bugs(code: str, tests: str, result: str, feedback: str, model: LLM) -> str:
+    prompt = FIX_BUG.format(completion=code, test_case=tests, result=result)
+    completion = model(prompt)
+    return preprocess_data(completion)
+
+
+class AgentCoder(Agent):
+    """AgentCoder is based off of the AgentCoder paper https://arxiv.org/abs/2312.13010
+    and it's open source code https://github.com/huangd1999/AgentCoder with some key
+    differences. AgentCoder comprises of 3 components: a coder agent, a tester agent,
+    and an executor. The tester agents writes code to test the code written by the coder
+    agent, but in our case because we are solving a vision task it's difficult to write
+    testing code. We instead have the tester agent write code to visualize the output
+    of the code written by the coder agent. If the code fails, we pass it back to the
+    coder agent to fix the bug, if it succeeds we pass it to a visual tester agent, which
+    is an LMM model like GPT4V, to visually inspect the output and make sure it looks
+    good."""
+
+    def __init__(
+        self,
+        coder_agent: Optional[LLM] = None,
+        tester_agent: Optional[LLM] = None,
+        visual_tester_agent: Optional[LMM] = None,
+        verbose: bool = False,
+    ) -> None:
+        self.coder_agent = (
+            OpenAILLM(temperature=0.1) if coder_agent is None else coder_agent
+        )
+        self.tester_agent = (
+            OpenAILLM(temperature=0.1) if tester_agent is None else tester_agent
+        )
+        self.visual_tester_agent = (
+            OpenAILMM(temperature=0.1, json_mode=True)
+            if visual_tester_agent is None
+            else visual_tester_agent
+        )
+        self.max_turns = 3
+        if verbose:
+            _LOGGER.setLevel(logging.INFO)
+
+    def __call__(
+        self,
+        input: Union[List[Dict[str, str]], str],
+        image: Optional[Union[str, Path]] = None,
+    ) -> str:
+        if isinstance(input, str):
+            input = [{"role": "user", "content": input}]
+        return self.chat(input, image)
+
+    def chat(
+        self,
+        input: List[Dict[str, str]],
+        image: Optional[Union[str, Path]] = None,
+    ) -> str:
+        question = input[0]["content"]
+        if image:
+            question += f" Input file path: {os.path.abspath(image)}"
+
+        code = ""
+        feedback = ""
+        for _ in range(self.max_turns):
+            code = write_program(question, feedback, self.coder_agent)
+            _LOGGER.info(f"code:\n{code}")
+            debug = write_debug(question, code, feedback, self.tester_agent)
```

Comment:
What happens for tasks where we don't have visualizations ? Like Image to text models. It just executes the code ?